### PR TITLE
coprocessor/dag/expr: add Date for built-in function

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: hawkingrei <hawking.rei@gmail.com>

## What have you changed? (mandatory)

add Date for mysql built-in function. see https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date

## What are the type of the changes? (mandatory)
featurn

## How has this PR been tested? (mandatory)
unittest

Signed-off-by: hawkingrei <hawking.rei@gmail.com>

### CODE REVIEW

**Comment by hawkingrei** [comment]
PLAT @AndreMouche

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 1]
```diff
@@ -31,6 +32,24 @@ impl ScalarFunc {
         let res = t.date_format(format_mask_str)?;
         Ok(Some(Cow::Owned(res.into_bytes())))
     }
+
+    #[inline]
+    pub fn eval_data<'a, 'b: 'a>(
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could we rename it to `date` instead?

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 2]
```diff
@@ -31,6 +32,24 @@ impl ScalarFunc {
...
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        if t.invalid_zero() {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Seems we need to check `is_zero` here according to https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L269

File: src/coprocessor/dag/expr/scalar_function.rs [hunk 3]
```diff
@@ -856,7 +856,7 @@ dispatch_call! {
         CastTimeAsTime => cast_time_as_time,
         CastDurationAsTime => cast_duration_as_time,
         CastJsonAsTime => cast_json_as_time,
-
+        Date => eval_data,
```
Comment (breeswish) [inline src/coprocessor/dag/expr/scalar_function.rs]: Name it as `date` like others should be better.

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 4]
```diff
@@ -91,4 +110,21 @@ mod test {
             assert_eq!(got, Datum::Bytes(exp.to_string().into_bytes()));
         }
     }
+
+    #[test]
+    fn test_date() {
+        let tests = vec![
+            ("2011-11-11", "2011-11-11"),
+            ("2011-11-11 10:10:10", "2011-11-11"),
+        ];
```
Comment (breeswish) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could you also add tests for zero dates and invalid dates?

**Comment by hawkingrei** [comment]
I will add the test for zero dates and invalid dates as soon as possible

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 5]
```diff
@@ -31,11 +32,32 @@ impl ScalarFunc {
...
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        if t.invalid_zero() {
+            return Err(box_err!("Incorrect datetime value: '{}'", t));
+        }
+        if t.is_zero() {
+            return Err(box_err!("Incorrect datetime value: '{}'", t));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could we define a similar error with the same error code and the similar message like tidb?

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 2 above)
Comment (hawkingrei) [reply src/coprocessor/dag/expr/builtin_time.rs]: Done

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 4 above)
Comment (hawkingrei) [reply src/coprocessor/dag/expr/builtin_time.rs]: Done

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 6]
```diff
@@ -31,6 +32,27 @@ impl ScalarFunc {
...
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let t = try_opt!(self.children[0].eval_time(ctx, row));
+        if t.invalid_zero() {
+            return Err(box_err!("Incorrect datetime value: '{}'", t));
+        }
+        if t.is_zero() {
+            return Err(box_err!("Datetime value is zero timestamp: '{}'", t));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: I think we need to return an error with the same error code as tidb did, you can define it similar to https://github.com/tikv/tikv/blob/master/src/coprocessor/codec/error.rs#L69
Also you need to handle the error with ctx like tidb did in `handleInvalidTimeError`

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 7]
```diff
@@ -12,9 +12,19 @@
 // limitations under the License.
 
 use super::{EvalContext, Result, ScalarFunc};
+use coprocessor::codec::error::Error;
+use coprocessor::codec::mysql::{self, Time};
 use coprocessor::codec::Datum;
 use std::borrow::Cow;
 
+fn handle_invalid_time_error(ctx: &mut EvalContext, err: Error) -> Option<Error> {
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: Could we define this function in `EvalContext`? like `ctx.handle_division_by_zero` did（https://github.com/tikv/tikv/blob/master/src/coprocessor/dag/expr/ctx.rs#L228）

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 8]
```diff
@@ -31,14 +41,38 @@ impl ScalarFunc {
...
         Ok(Some(Cow::Owned(res.into_bytes())))
     }
+
+    #[inline]
+    pub fn date<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let t = try_opt!(self.children[0].eval_time(ctx, row));
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: It seems we need to handle the error as invalid time error when eval_time failed according to tidb https://github.com/pingcap/tidb/blob/master/expression/builtin_time.go#L272

File: src/coprocessor/dag/expr/builtin_time.rs [hunk 9]
```diff
@@ -31,14 +41,38 @@ impl ScalarFunc {
...
+        if t.is_zero() {
+            match handle_invalid_time_error(ctx, Error::incorrect_datetime_value(&format!("{}", t)))
+            {
+                Some(err) => return Err(err),
+                None => return Ok(None),
+            }
+        }
+        let res = Time::new(
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_time.rs]: could we use the function `set_tp` in `Time` here?

**Comment by hawkingrei** [comment]
I will refactor date_format and date using handleInvalidTimeError instead of try_opt!.

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 7 above)
Comment (hawkingrei) [reply src/coprocessor/dag/expr/builtin_time.rs]: Done

File: src/coprocessor/dag/expr/builtin_time.rs (hunk 8 above)
Comment (hawkingrei) [reply src/coprocessor/dag/expr/builtin_time.rs]: Done

**Pull request closed without merging**